### PR TITLE
fix(s3): verify SigV4 signature on presigned URLs when enforce-auth is enabled

### LINKED ISSUE #1841: [BUG] S3: presigned URL signature is not verified even with enforce-auth enabled
## Service

S3

## AWS API Action

GetObject, PutObject (any presigned URL access)

## Expected behavior

- enforce-auth enabled + tampered X-Amz-Signature → 403 SignatureDoesNotMatch

## Actual behavior

- tampered signature still returns 200 OK
- only expiration and access-key-existence are checked
- signature value is never verified

## Reproduction
```
docker run -d -p 4566:4566 -e FLOCI_SERVICES_S3_ENFORCE_AUTH=true floci/floci:latest

aws --endpoint-url=http://localhost:4566 s3 mb s3://test-bucket
echo "secret" | aws --endpoint-url=http://localhost:4566 s3 cp - s3://test-bucket/file.txt

URL=$(aws --endpoint-url=http://localhost:4566 s3 presign s3://test-bucket/file.txt --expires-in 60)
curl "$URL"                                    # 200 (correct)
curl "$(echo "$URL" | sed 's/.$/0/')"          # 200 (should be 403)
```

## Root cause

- PreSignedUrlFilter only verifies signatures when validate-signatures is enabled (default false)
- validate-signatures uses a custom (non-SigV4) algorithm → rejects all AWS SDK URLs
- enforce-auth gates access-key existence only, never verifies actual signature

## Environment

- floci 1.5.32
- AWS CLI 2.x / AWS SDK for Java 2.31.1

PR: #1842

### PR DESCRIPTION
**Pull request by ecsimsw**
## Summary
- Even with `enforce-auth` enabled, presigned URLs with a tampered `X-Amz-Signature` still return 200. 
- Only expiration and access-key existence are checked, the signaturevalue itself is never verified.
- This PR adds AWS SigV4 signature verification in `PreSignedUrlFilter` so that tampered presigned URLs are rejected with 403 `SignatureDoesNotMatch`, matching real AWS behavior.

Closes #1841

## Why

- Without signature verification, `enforce-auth` cannot prevent access via a modified presigned URL. 
- Any client that knows the URL pattern can change the path or expiry and still access the object, effectively making private content public.
- Additionally, users who write tests asserting that a tampered signature returns 403 (as real AWS does) will see those tests fail only on floci, not on AWS.

## Note
- X-Amz-Algorithm and X-Amz-Expires validations are applied unconditionally, independent of the enforce-auth flag. 
- This matches the behavior of other S3-compatible implementations (e.g. MinIO).

## What changed

- `PreSignedUrlFilter`: when `enforce-auth` is enabled, recompute the SigV4 signature and compare it against the request's signature value
- S3 auth enforcement integration tests: replace hardcoded fake signatures with real SigV4-computed signatures so tests pass with verification enabled
- Follows the code style and structure of the existing RDS and ElastiCache SigV4 validators
- The existing `validate-signatures` flag behavior is preserved for backward compatibility
- Payload hash reads from `X-Amz-Content-Sha256` (header → query param → `UNSIGNED-PAYLOAD` fallback) instead of hardcoding
- `X-Amz-Expires` validated to 1–604800 range, returns `AuthorizationQueryParametersError` outside that range
- `X-Amz-Algorithm` value checked early, rejects unsupported values with `AuthorizationQueryParametersError`

## Note
- SigV4 verification only runs when `enforce-auth` is enabled. When disabled (default), behavior is unchanged. 
- The SigV4 utility methods are now duplicated across RDS, ElastiCache, and S3. 
- I think extracting them into a shared utility in `core/common/` would benefit code consistency and future contributors. 
- I kept the same duplicated form to match the existing style, but if you agree with extracting, I'm happy to add that work.

## Type of change

- [x] Bug fix (`fix:`)

## AWS Compatibility

Tested with AWS CLI `aws s3 presign` and AWS SDK for Java `S3Presigner`:
- Valid signature returns 200
- Tampered signature returns 403 SignatureDoesNotMatch
- Expired URL returns 403 AccessDenied
- Unknown access key returns 403 InvalidAccessKeyId

## Checklist

- [x] `./mvnw test` passes locally
- [x] New or updated integration test added
- [x] Commit messages follow Conventional Commits

### CODE REVIEW

File: src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java [hunk 1]
```diff
@@ -76,6 +99,122 @@ public void filter(ContainerRequestContext requestContext) {
...
+            String accessKeyId = credParts[0];
+            String date = credParts[1];
+            String region = credParts[2];
+            String service = credParts[3];
+            String credentialScope = date + "/" + region + "/" + service + "/aws4_request";
+
+            String secretKey = resolveSecretKey(accessKeyId);
+            if (secretKey == null) {
+                return false;
+            }
```
[the remark this answers is not in the document]
Comment (ecsimsw) [reply src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java]: Fixed in the latest push. 
Unknown access keys now return InvalidAccessKeyId before signature verification, and the test assertion has been restored to containsString("InvalidAccessKeyId").

File: src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java [hunk 2]
```diff
@@ -56,8 +74,25 @@ public void filter(ContainerRequestContext requestContext) {
...
+            String decodedCredential = URLDecoder.decode(credential, StandardCharsets.UTF_8);
+            String[] credParts = decodedCredential.split("/");
+            String accessKeyId = credParts.length >= 5 ? credParts[0] : null;
+
+            String secretKey = resolveSecretKey(accessKeyId);
+            if (secretKey == null) {
+                requestContext.abortWith(errorResponse(403, "InvalidAccessKeyId",
+                        "The AWS Access Key Id you provided does not exist in our records."));
+                return;
+            }
```
[the remark this answers is not in the document]
Comment (ecsimsw) [reply src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java:121]: Fixed in the latest push.
Added an early return with InvalidAccessKeyId when credParts.length < 5, before accessing credParts[0].

File: src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java [hunk 3]
```diff
@@ -76,6 +99,122 @@ public void filter(ContainerRequestContext requestContext) {
...
+
+    private String resolveSecretKey(String accessKeyId) {
+        if (LEGACY_ACCESS_KEY_ID.equals(accessKeyId)) {
+            return LEGACY_SECRET_KEY;
+        }
+        if (iamService != null) {
+            return iamService.findSecretKey(accessKeyId).orElse(null);
+        }
+        return null;
+    }
```
[the remark this answers is not in the document]
Comment (ecsimsw) [reply src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java:224]: I looked into this. 

floci uses test/test as its default credential, and the same pattern is already in S3Service#isKnownAccessKey (LEGACY_ACCESS_KEY_ID = "test"). 

That handling was introduced in #1689 — the S3 auth enforcement feature itself — so accepting test under enforce-auth is by design of that feature, not something this PR added.

Conversely, removing test/test would make the default authentication that works everywhere else in S3 stop working only in this PR's presigned-URL verification, which is out of sync with floci's existing behavior.

[the remark this answers is not in the document]
Comment (ecsimsw) [reply src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java:224]: Thanks for confirming

**Comment by snazy** [comment]
Could we add the actual regression case here? Generate standard presigned URLs for two objects, verify both original URLs succeed, then combine object A’s path with object B’s query and assert `403 SignatureDoesNotMatch`.
The current tests only exercise a valid signature produced by a helper that duplicates the production algorithm, so CI does not verify the defect this PR is intended to fix. Prefer using `S3Presigner` in the existing Java compatibility suite.

This does not fully implement SigV4 canonical query construction. It retains the raw query pairs and sorts only by parameter name, whereas SigV4 requires URI-encoding each name and value and sorting by encoded name and then encoded value.
Could we canonicalize parsed name/value pairs according to the AWS rules and cover reordered parameters, reserved characters, empty values, and duplicate names? As written, some AWS-valid URLs may be rejected, while noncanonical signatures that AWS rejects may be accepted.

**Comment by snazy** [comment]
Thanks for the follow-up commits. I reviewed both `2677002` and `d379d83`. They improve the PR, but they do not yet fully address my previous comment. To avoid any ambiguity, I am asking for two separate things here, and I consider both necessary before this concern is resolved.

### 1. Add the actual path/query-transplant regression using an independent signer

The new `presignedRequestWithTamperedSignatureIsRejected` test changes the final character of `X-Amz-Signature`. That is useful coverage: it proves that the supplied signature value is now compared against a calculated value.

However, it does not test the reported invariant: a signature generated for one S3 resource must not authorize a request to another resource.

It also continues to generate the signature through the test-local `presignedSignature` helper. That helper reproduces the same canonical-request and signing logic as the production implementation. If both implementations make the same canonicalization mistake, the positive and negative tests can still pass while standard AWS SDK-generated URLs fail.

Please add a regression test with these properties:

1. Run Floci with `floci.services.s3.enforce-auth=true`, so the standard SigV4 verification branch is definitely exercised. The test must not silently skip when enforcement is disabled.
2. Create one bucket with two different object keys and preferably different contents.
3. Use the AWS SDK for Java `S3Presigner` with credentials accepted by that Floci instance to generate a standard presigned GET URL for each object.
4. Verify that both original SDK-generated URLs succeed. This is important because it independently proves that Floci accepts the SDK’s SigV4 construction.
5. Construct a transplanted URL using the raw path from object A and the complete raw query from object B. Please preserve the raw encoded path and query rather than decoding and rebuilding them.
6. Send a GET request to that transplanted URL.
7. Assert HTTP 403 and the S3 XML error code `SignatureDoesNotMatch`.

The existing tampered-last-character test can remain as additional coverage, but it does not replace this test. Likewise, another test using the hand-written `presignedSignature` helper would not provide the independent AWS SDK compatibility evidence requested here.

The existing `compatibility-tests/sdk-test-java` module already has `S3Presigner` available, so that looks like the natural place for this regression if enabling strict S3 authentication for the test can be made deterministic.

### 2. Add focused coverage for the new canonical-query implementation

The new implementation is moving in the correct direction: it re-encodes decoded parameter names and values, sorts by encoded name and then encoded value, renders `name=value`, and excludes `X-Amz-Signature`.

But the commit changes security-sensitive canonicalization behavior without adding tests for the cases that motivated the change. The existing tests only contain the standard, uniquely named `X-Amz-*` parameters in already sorted order. They therefore do not exercise the new encoding, value sorting, empty-value, or duplicate-name branches.

Please add focused tests covering at least:

- Query parameters arriving in a different order while still validating against the same canonical ordering.
- Reserved characters and spaces being encoded according to SigV4 rules: uppercase percent escapes, `%20` for spaces rather than `+`, and encoding of characters such as `/`, `%`, and `+`.
- Non-ASCII UTF-8 values, so encoding occurs byte-by-byte rather than character-by-character.
- An empty value being represented canonically as `name=`.
- Duplicate parameter names being sorted by their encoded values.
- `X-Amz-Signature` being excluded while all other parameters remain part of the canonical query.

These can be unit tests around an extracted/package-private canonical-query function or integration tests through the request filter. The important part is that they exercise the framework’s actual decoded query representation and compare the resulting canonical form or verification result against independently known expectations.

To be completely explicit about the acceptance criteria: I will consider my earlier comment resolved when the PR includes both the SDK-generated path/query-transplant regression and meaningful coverage of the canonical-query edge cases above, with the relevant tests running successfully under S3 authentication enforcement.

**Comment by ecsimsw** [comment]
@snazy 
Thanks for the thorough review. 
I split the coverage into unit and integration tests.

**Unit tests** 
- extracted the canonical query construction into a package-private method
- Parameters arriving in a different order still sort by encoded name and value
- Reserved characters and spaces encoded per SigV4 (`%20`, `/`, `%`, `+` → uppercase `%XX`)
- Non-ASCII UTF-8 encoded byte-by-byte
- Empty value → `name=`
- Duplicate names sorted by their encoded values
- `X-Amz-Signature` excluded while every other parameter remains

**Integration tests** 
- enforce-auth=true, AWS SDK `S3Presigner`
- Two valid SDK URLs succeed + transplanting object A's path onto object B's query → 403 `SignatureDoesNotMatch`
- Reordered query still verifies
- Query with special characters (SDK-signed) is accepted
- Expired URL is rejected
- URL signed with an IAM-issued access key is accepted
- Credential with fewer than 5 scope parts is rejected

### Notes
- I placed this in the main integration tests rather than compat `sdk-test-java`. My understanding is that the compat suite focuses on compatibility with real SDK/CLI clients, while this PR verifies a functional behavior, so an integration test felt like a natural fit.
- The integration test can enable enforce-auth for just this class via `@TestProfile`, with no infrastructure changes, so it applied cleanly.
- Added `aws-sdk s3` as a test dependency.

**Comment by snazy** [comment]
Thanks for the detailed follow-up. I reviewed the added canonical-query unit tests and SDK-presigner integration coverage. They address both parts of my earlier comment.

The SDK test now exercises the actual regression under `floci.services.s3.enforce-auth=true`: both independently presigned object URLs succeed, while combining object A’s raw path with object B’s complete raw query is rejected with `403 SignatureDoesNotMatch`.

The focused unit tests cover query ordering, SigV4 encoding, UTF-8, empty values, duplicate names, and exclusion of `X-Amz-Signature`.

I am not asking for further changes to those two items.

One separate, non-blocking compatibility edge remains in the original signature-verification implementation: canonical values for signed headers other than host are only processed with `trim()`, without collapsing sequential whitespace as AWS signers do during canonicalization. That could reject an otherwise valid presigned request containing such a header.

I figured out that inactive IAM access keys and expired temporary credentials are happily accepted as valid, which predates this PR, so I opened #1934.

**Comment by ecsimsw** [comment]
@snazy 
Thanks again for the detailed review. 
It helped a lot, and I learned a great deal from you. 
In particular I revisited how to write a regression test properly. 

The sequential-whitespace handling for non-host signed headers isn't part of #1934, 
so I've handled it here in this PR (46fe2dd).

**Comment by snazy** [comment]
> @snazy Thanks again for the detailed review. It helped a lot, and I learned a great deal from you. In particular I revisited how to write a regression test properly.

YW!
I've been hitting the same issue for a project and it was great to see that it's already been tackled!

**Comment by hectorvent** [comment]
Hi @ecsimsw, 

Thanks for the PR. Please take a look at the [CONTRIBUTING.md](https://github.com/floci-io/floci/blob/main/CONTRIBUTING.md). The AWS dependency are already on the compatibility-tests java project.

**Comment by ecsimsw** [comment]
@hectorvent 
Thanks for the feedback.
I moved the SDK integration test to compatibility-tests/sdk-test-java and removed the aws-sdk:s3 dependency from the main pom.xml 

This test requires enforce-auth=true, but enabling it globally would break other S3 tests. 
Following the pattern in S3AuthEnforcementCompatibilityTest, tests are skipped via `Assumptions.assumeTrue` when enforcement is not active. 
They only run when Floci is started with `enforce-auth=true`.

File: src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java [hunk 4]
```diff
@@ -76,6 +121,152 @@ public void filter(ContainerRequestContext requestContext) {
...
+
+            // Canonical request
+            String path = requestContext.getUriInfo().getRequestUri().getRawPath();
+            String canonicalQueryString = buildCanonicalQueryString(queryParams);
+            String canonicalRequest = requestContext.getMethod() + "\n"
+                    + path + "\n"
+                    + canonicalQueryString + "\n"
+                    + canonicalHeaders + "\n"
+                    + signedHeaders + "\n"
+                    + "UNSIGNED-PAYLOAD";
```
Comment (abanna) [inline src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java]: Good to see this verified against SDK-generated URLs rather than only against Floci's own signer — that's the test that actually proves the canonicalization, and `MessageDigest.isEqual` for the comparison is the right call.

The payload hash is hardcoded here, though, and that will falsely reject some legitimately signed URLs. In SigV4 the final line of the canonical request is the hashed payload, which for a presigned request is whatever the signer put in `X-Amz-Content-Sha256` — `UNSIGNED-PAYLOAD` is only the *common* value, not the required one. A client that presigns with an explicit content hash (or `STREAMING-UNSIGNED-PAYLOAD-TRAILER`) signs `X-Amz-Content-Sha256` as a signed header, so it already flows into `canonicalHeaders` two lines up — but the hash position stays `UNSIGNED-PAYLOAD`, the computed signature diverges, and the request gets a 403 `SignatureDoesNotMatch` despite being correctly signed.

Taking it from the header when present would cover both:

```java
String payloadHash = requestContext.getHeaderString("x-amz-content-sha256");
if (payloadHash == null) {
    payloadHash = queryParams.getFirst("X-Amz-Content-Sha256");
}
if (payloadHash == null) {
    payloadHash = "UNSIGNED-PAYLOAD";
}
```

Separately, on the parse above at line 68: `X-Amz-Expires` is accepted with no range check. AWS constrains it to 1–604800 and rejects anything outside that with `AuthorizationQueryParametersError`, so `?X-Amz-Expires=999999999` currently produces a presigned URL that Floci treats as valid essentially forever. Since the point of this PR is to make `enforce-auth` mean something, an unbounded expiry seems worth closing at the same time.

One smaller thing in the same area: `X-Amz-Algorithm` is only checked for presence at line 49, never for value, while `verifySigV4Signature` hardcodes `AWS4-HMAC-SHA256` in the string-to-sign. It isn't a bypass — a bogus algorithm just fails the signature — but the caller gets `SignatureDoesNotMatch` where AWS returns `AuthorizationQueryParametersError`, which is a confusing thing to debug against.

**Comment by hectorvent** [comment]
Thank you for taking on signature verification, which is fiddly to get right.

**(blocking)** How does this interact with `FLOCI_AUTH_VALIDATE_SIGNATURES`, which defaults to `false`? If presigned verification is now unconditional while normal request signing stays opt-in, that is a behavior change worth calling out in the description, since existing setups with slightly wrong presign code would start failing.

**(blocking)** The negative paths need pinning with tests: an expired URL, a tampered query parameter, and a signature computed over a different canonical request. For a change that decides whether requests are rejected, those three are the evidence that the verification is neither too strict nor too loose.

Both points are blockers for a security-adjacent change; the direction itself is right.

**Comment by ecsimsw** [comment]
@abanna 
Addressed all three:
1. Payload hash : reads X-Amz-Content-Sha256 from header/query param, falls back to UNSIGNED-PAYLOAD
2. Expires range : validates 1–604800, returns AuthorizationQueryParametersError
3. Algorithm check : rejects unsupported values early with AuthorizationQueryParametersError

Thanks for the thorough review.

**Comment by ecsimsw** [comment]
@hectorvent
Both blocking points are already covered:

#### SigV4 verification is not unconditional :
- it's gated behind [`isAuthEnforced()`](https://github.com/ecsimsw/floci/blob/f0652b59/src/main/java/io/github/hectorvent/floci/services/s3/PreSignedUrlFilter.java#L105), so it only runs when enforce-auth is enabled. 
- Default behavior (false) is unchanged.

#### All three negative paths have tests:
- expiredPresignedUrlIsRejected : sets X-Amz-Date to 2020, verifies 403 rejection                                              
- presignedRequestWithTamperedSignatureIsRejected : flips the last hex digit of the signature, verifies 403                  
  SignatureDoesNotMatch                                                                                                           
- sdkPresignedUrlIsAcceptedButNotValidForAnotherObject : signs URL for object-a, swaps path to object-b, verifies 403 from canonical request mismatch

Will update the PR description to make the enforce-auth relationship clearer.

**Review by abanna — APPROVED** [review]
Thank you for the careful S3 authentication work. I reviewed the current head and ran `PreSignedUrlFilterTest` plus `S3AuthEnforcementIntegrationTest`—all 35 tests passed, and the full CI set is green. The implementation includes AWS SDK-based presigning coverage and follows the repository’s contribution, testing, and commit conventions. I found no blocking issues. Nice work—approved.

**Comment by ecsimsw** [comment]
@abanna Thanks for the thorough review and approval!

**Comment by hectorvent** [comment]
Thank you, and I am sorry this sat over a week after your reply. Both of my blocking points are answered, and I followed them through the filter rather than take them on trust: the SigV4 path sits behind `isAuthEnforced()` with the old `validate-signatures` branch untouched underneath, so the default really is unchanged, and the description now says so. The negative paths are pinned too, including the tampered signature, expires out of range, and unsupported algorithm.

Consider both asks withdrawn. @abanna has already approved the whole thing, so I will not repeat that.

**(follow-up, separate PR)** The `X-Amz-Algorithm` and `X-Amz-Expires` checks sit above that gate, so they apply with enforce-auth off as well. The codes match what other S3 implementations return ([minio](https://github.com/minio/minio/blob/master/cmd/api-errors.go)), so I would keep the behaviour, just worth a line in the description.

**(follow-up, separate PR)** And yes please to your offer of pulling the shared SigV4 helpers into `core/common`.

**Comment by ecsimsw** [comment]
@hectorvent 
Thank you for the merge! 
Contributing to project I actively use in production made this especially meaningful.

I'll take on extracting the shared SigV4 helpers into `core/common/` as a follow-up PR.
I've also updated the description to note that theX-Amz-Algorithm and X-Amz-Expires validations run regardless of enforce-auth.
